### PR TITLE
feat(minor-axelarnet-gateway): call contract with token to nexus module

### PR DESCRIPTION
**Pull request by haiyizxx**
## Todos

- [ ] Unit tests
- [ ] Manual tests
- [ ] Documentation
- [ ] Connect epics/issues

### CODE REVIEW

File: packages/axelar-wasm-std/src/response.rs [hunk 1]
```diff
@@ -13,9 +13,9 @@ pub enum Error {
 
 /// Get a msg wrapped inside a `WasmMsg::Execute` from a `Response`.
 /// If there are no wasm messages or more than one message in the response, this returns an error.
-pub fn inspect_response_msg<T>(response: Response) -> Result<T, Error>
+pub fn inspect_response_msg<T, D>(response: Response<T>) -> Result<D, Error>
 where
-    T: DeserializeOwned,
+    D: DeserializeOwned,
```
Comment (milapsheth) [inline packages/axelar-wasm-std/src/response.rs:18]: Use `Empty` as the default similar to what `Response` does?
```suggestion
pub fn inspect_response_msg<T, D = Empty>(response: Response<D>) -> Result<T, Error>
where
    T: DeserializeOwned,
```

File: contracts/axelarnet-gateway/tests/utils/mod.rs [hunk 2]
```diff
@@ -2,11 +2,72 @@
...
 pub use execute::*;
 pub use instantiate::*;
+use serde_json::json;
 
 mod execute;
 mod instantiate;
 pub mod messages;
 pub mod params;
+
+pub fn mock_axelar_dependencies(
```
Comment (milapsheth) [inline contracts/axelarnet-gateway/tests/utils/mod.rs:22]: Move these to `instantiate` or create another module perhaps?

File: contracts/axelarnet-gateway/tests/utils/mod.rs [hunk 3]
```diff
@@ -2,11 +2,72 @@
...
+) -> OwnedDeps<MockStorage, MockApi, MockQuerier<AxelarQueryMsg>, AxelarQueryMsg> {
+    OwnedDeps {
+        storage: MockStorage::default(),
+        api: MockApi::default(),
+        querier: MockQuerier::<AxelarQueryMsg>::new(&[("contract", &[])]),
+        custom_query_type: PhantomData,
+    }
+}
+
+pub fn axelar_query_handler(
```
Comment (milapsheth) [inline contracts/axelarnet-gateway/tests/utils/mod.rs:32]: Move this to `query.rs` maybe?

File: contracts/axelarnet-gateway/src/contract/execute.rs [hunk 4]
```diff
@@ -74,30 +76,28 @@ impl CallContractData {
...
 }
 
+enum RoutingDestination {
+    Nexus,
+    Router,
+}
+
+type Result<T> = error_stack::Result<T, Error>;
+
 pub fn call_contract(
```
Comment (milapsheth) [inline contracts/axelarnet-gateway/src/contract/execute.rs:86]: Update the doc in `msg.rs` for CallContract to document behaviour differences depending on chain type / attached token

File: contracts/axelarnet-gateway/src/contract/execute.rs [hunk 5]
```diff
@@ -253,3 +254,56 @@ fn try_load_executable_msg(
...
+        .map(|token| BankMsg::Send {
+            to_address: nexus.to_string(),
+            amount: vec![token],
+        })
+        .map(Into::into)
+        .into_iter()
+        .chain(iter::once(client.route_message(msg)))
+        .collect::<Vec<_>>()
+        .then(|msgs| Response::new().add_messages(msgs))
+        .then(Ok)
```
Comment (milapsheth) [inline contracts/axelarnet-gateway/src/contract/execute.rs:308]: Don't really find this very readable imo. Maybe something more imperative but simpler instead?

```suggestion
    let mut response_msgs: Vec<CosmosMsg<nexus::execute::Message>> = vec![];

    if let Some(token) = token.clone() {
        response_msgs.push(BankMsg::Send {
            to_address: nexus.to_string(),
            amount: vec![token],
        }.into())
    };

    response_msgs.push(client.route_message((msg, token).into()));
    Response::new().add_messages(response_msgs).then(Ok)
```

File: contracts/axelarnet-gateway/src/contract/execute.rs [hunk 6]
```diff
@@ -53,6 +53,8 @@ pub enum Error {
     NonceOverflow,
     #[error("invalid token received")]
     InvalidToken,
+    #[error("invalid routing destination")]
+    RoutingDestination,
```
Comment (milapsheth) [inline contracts/axelarnet-gateway/src/contract/execute.rs:57]:
```suggestion
    InvalidRoutingDestination,
```

File: contracts/axelarnet-gateway/tests/execute.rs [hunk 7]
```diff
@@ -171,18 +165,18 @@ fn route_to_router_after_contract_call_with_tempered_data_fails() {
...
         .querier
-        .with_custom_handler(reply_with_tx_hash_and_nonce(tx_hash, nonce));
+        .with_custom_handler(axelar_query_handler(tx_hash, nonce, false));
 
     let destination_chain = "destination-chain".parse().unwrap();
     let destination_address = "destination-address".parse().unwrap();
     let payload = vec![1, 2, 3].into();
 
-    utils::instantiate_contract(deps.as_mut()).unwrap();
+    utils::instantiate_contract(emptying_deps_mut(&mut deps.as_mut())).unwrap();
```
Comment (milapsheth) [inline contracts/axelarnet-gateway/tests/execute.rs:177]: Assign a variable to the emptying deps instead of having to convert it everywhere

**Review by milapsheth — APPROVED** [review]

File: packages/axelar-wasm-std/src/response.rs (hunk 1 above)
Comment (haiyizxx) [reply packages/axelar-wasm-std/src/response.rs:18]: na, default is not allowed for function definations

File: contracts/axelarnet-gateway/src/msg.rs [hunk 8]
```diff
@@ -11,8 +11,8 @@ pub struct InstantiateMsg {
     pub chain_name: ChainName,
     /// Address of the router contract on axelar.
     pub router_address: String,
-    /// Address of the nexus gateway contract on axelar.
-    pub nexus_gateway: String,
+    /// Address of the nexus module on axelar.
```
Comment (fish-sammy) [inline contracts/axelarnet-gateway/src/msg.rs:14]:
```suggestion
    /// Address of the nexus module account on axelar.
```

File: contracts/axelarnet-gateway/tests/execute.rs (hunk 7 above)
Comment (haiyizxx) [reply contracts/axelarnet-gateway/tests/execute.rs:177]: I also don't like that, but the function consumes the deps mut

File: contracts/axelarnet-gateway/src/contract/execute.rs (hunk 5 above)
Comment (fish-sammy) [reply contracts/axelarnet-gateway/src/contract/execute.rs:308]: I like the current implementation better and more elegant.

**Pull request merged**